source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
+const keep_ops = { // Operations for keeping or dropping rolls of a result set
+	KEEP: 'k',
+	KEEP_HIGHEST: 'kh',
+	DROP_LOWEST: 'dl',
+	DROP: 'd',
+	DROP_HIGHEST: 'dh',
+	KEEP_LOWEST: 'kl'
+}
+
+class ResultSet {
+
+	constructor() {}
+
+	static #surroundedByParens(str, symbolPos) {
+		var rightParensUnenclosed = 0,
+			leftParensUnenclosed = 0;
+		var leftSide = str.slice(0,symbolPos),
+			rightSide = str.slice(symbolPos+1,str.length-1);
+		for (var i = leftSide.length - 1; i >= 0 ; i-=1) {
+			if (leftSide[i] == ')' && leftParensUnenclosed <= 0) {
+				leftParensUnenclosed -= 1;
+			} else if (leftSide[i] == '(') {
+				leftParensUnenclosed += 1;
+			}
+		}
+		for (var i = 0; i < rightSide.length; i+=1) {
+			if (rightSide[i] == '(' && rightParensUnenclosed <= 0) {
+				rightParensUnenclosed -= 1;
+			} else if (rightSide[i] == ')') {
+				rightParensUnenclosed += 1;
+			}
+		}
+		return rightParensUnenclosed > 0 && leftParensUnenclosed > 0;
+	}
+
+	static #parseStringList(str) {
+		let strList = [];
+		let newEntry = "";
+		let startQuote = false;
+		let endQuote = false;
+		let spaceCount = 0;
+		for (var i = 0; i < str.length; i++) {
+			if (
+				(str[i] == '[' && i == 0) ||
+				(!(startQuote && endQuote) && str[i] == ']' && i == str.length-1) ||
+				(startQuote && endQuote && str[i] == ' ')
+				) {
+					if (str[i] == ' ') {
+						spaceCount++;
+					}
+				continue;
+			} else if (startQuote && endQuote && str[i] == ']' && i == str.length-1) {
+				strList.push(newEntry.trim());
+				newEntry = "";
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && endQuote && str[i] === ',') {
+				strList.push(newEntry.trim());
+				newEntry = "";
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && endQuote) {
+				newEntry = `"${newEntry}"${' '.repeat(spaceCount)}${str[i]}`;
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && !endQuote && str[i] == "\"") {
+				endQuote = true;
+				continue;
+			} else if (!startQuote && !endQuote && str[i] == "\"") {
+				startQuote = true;
+				continue;
+			} else if (!startQuote && str[i] == ",") {
+				strList.push(newEntry.trim());
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				newEntry = "";
+				continue;
+			}
+			newEntry = newEntry + str[i];
+		}
+		if (startQuote && !endQuote) {
+			throw `Mismatched quotation marks in list:\n${str}`;
+		}
+		let lastEntry = newEntry.trim();
+		if (lastEntry) {
+			strList.push(lastEntry);
+		}
+		return strList;
+	}
+
+	static parse(value) { // Calls the right ResultSet constructor for the given string
+		var parsingString = value.slice(0); // Copy the string
+		parsingString = parsingString.replace(/(?<!\[[^\]]*)\s(?![^\[]*\])/g,''); //Remove all white space
+		parsingString = parsingString.replace(/\+\s*\-/g,'-'); //Simplify adding a negative to subtraction

```

Comment:
Does this regexp need the `\s`? White space should have been removed in the line above, right?

Reply:
L198 replaces subtraction with addition, is this even necessary?